src/Storages/StorageMongoDB.cpp

```diff
@@ -56,257 +59,339 @@ StorageMongoDB::StorageMongoDB(
-            document.add(name, static_cast<Float64>(column.getFloat32(idx)));
-        else if (which.isFloat64())
-            document.add(name, column.getFloat64(idx));
-        else if (which.isDate())
-            document.add(name, Poco::Timestamp(DateLUT::instance().fromDayNum(DayNum(column.getUInt(idx))) * 1000000));
-        else if (which.isDateTime())
-            document.add(name, Poco::Timestamp(column.getUInt(idx) * 1000000));
-        else
+        auto arr = array();
+        for (const auto & child : explist->children)
         {
-            WriteBufferFromOwnString ostr;
-            data_type.getDefaultSerialization()->serializeText(column, idx, ostr, FormatSettings{});
-            document.add(name, ostr.str());
+            if (const auto & child_func = child->as<ASTFunction>())
+                arr.append(visitWhereFunction(context, child_func));
+            else
+                throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Only constant expressions are supported in WHERE section");
         }
-    }
-
-    String collection_name;
-    String db_name;
-    StorageMetadataPtr metadata_snapshot;
-    std::shared_ptr<Poco::MongoDB::Connection> connection;
 
-    const bool is_wire_protocol_old;
-};
-
-
-Pipe StorageMongoDB::read(
-    const Names & column_names,
-    const StorageSnapshotPtr & storage_snapshot,
-    SelectQueryInfo & /*query_info*/,
-    ContextPtr /*context*/,
-    QueryProcessingStage::Enum /*processed_stage*/,
-    size_t max_block_size,
-    size_t /*num_streams*/)
-{
-    connectIfNotConnected();
-
-    storage_snapshot->check(column_names);
-
-    Block sample_block;
-    for (const String & column_name : column_names)
-    {
-        auto column_data = storage_snapshot->metadata->getColumns().getPhysical(column_name);
-        sample_block.insert({ column_data.type, column_data.name });
+        return make_document(kvp(func_name, std::move(arr)));
     }
 
-    return Pipe(std::make_shared<MongoDBSource>(connection, database_name, collection_name, Poco::MongoDB::Document{}, sample_block, max_block_size));
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Only constant expressions are supported in WHERE section");
 }
 
-SinkToStoragePtr StorageMongoDB::write(const ASTPtr & /* query */, const StorageMetadataPtr & metadata_snapshot, ContextPtr /* context */, bool /*async_insert*/)
+bsoncxx::document::value StorageMongoDB::buildMongoDBQuery(ContextPtr context, mongocxx::options::find & options, const SelectQueryInfo & query, const Block & sample_block)
 {
-    connectIfNotConnected();
-    return std::make_shared<StorageMongoDBSink>(collection_name, database_name, metadata_snapshot, connection);
-}
-
-StorageMongoDB::Configuration StorageMongoDB::getConfiguration(ASTs engine_args, ContextPtr context)
-{
-    Configuration configuration;
-
-    if (auto named_collection = tryGetNamedCollectionWithOverrides(engine_args, context))
+    auto & query_tree = query.query_tree->as<QueryNode &>();
```

Comment:
How we should handle this case?

Reply:
> @allmazz One less preferred option is to implement the same logic on top of the AST for the old analyzer. However, this is not encouraged since the old analyzer is deprecated and we generally avoid implementing new features there. Moreover, almost all tests now use the new analyzer by default, so the new feature in old infrastructure would be less tested.
> 
> Alternatively, we could keep the old implementation as it is and continue using it for the old analyzer. But it's not the only reason for retain Poco based implementation, it'll also allow smother transition for users in the next release. We will get rid of old implementation when new implementation is tested on real scenarios.

I decided to implement query building on top of query tree, and disable this feature when analyzed disabled.
Also, a setting added for enable the legacy implementation.